aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
-  bool missing_output_dtypes = false;
-  common_dtype_ = dtype();
-  for (auto& op : operands_) {
-    if (!op.tensor.defined() && !op.is_type_defined()) {
-      missing_dtypes = true;
-      if (op.is_output) {
-        missing_output_dtypes = true;
+    // Determines if there are varying input dtypes
+    // NOTE: the common dtype is set to the first defined input dtype observed
+    if (!op.is_output && op.target_dtype != common_dtype_) {
+      if (common_dtype_ == ScalarType::Undefined) {
+        common_dtype_ = op.target_dtype;
+      } else {
+        has_different_input_dtypes = true;
       }
     }
   }
 
-  if (common_dtype_strategy_ == CommonDTypeStrategy::PROMOTE_INPUTS) {
-    TORCH_CHECK(!missing_output_dtypes, "unable to compute and promote common dtype based only on inputs if there are missing dtypes for outputs");
-  }
+  // Checks that all inputs (and defined outputs) have the same dtype, if requested
+  if (check_all_same_dtype_) {
+    for (auto& op : operands_) {
+      if (!op.is_type_defined() || !op.tensor.defined()) {
+        continue;
+      }
 
-  bool compute_common_dtype = (common_dtype_strategy_ != CommonDTypeStrategy::NONE);
-  bool compute_common_dtype_only_for_inputs = (common_dtype_strategy_ == CommonDTypeStrategy::PROMOTE_INPUTS);
+      TORCH_CHECK(op.target_dtype == common_dtype_,
+                  "Found dtype ", op.target_dtype, " but expected ", common_dtype_);
+    }
+  }
 
-  bool may_have_differing_types = true;
-  bool common_device_is_cuda = false;
+  // Short-circuits if no additional work required
+  if (!has_undefined_outputs && !check_all_same_device_ &&
+      !promote_inputs_to_common_dtype_ && !cast_common_dtype_to_outputs_ &&
+      !enforce_safe_casting_to_output_) {
+    return;
+  }
 
-  if (missing_dtypes || compute_common_dtype) {
-    auto operands = compute_common_dtype_only_for_inputs ? at::ArrayRef<OperandInfo>(operands_).slice(noutputs()) : operands_;
-    auto common_type = compute_common_type_(operands);
-    auto common_device = std::get<0>(common_type);
-    common_device_is_cuda = common_device.is_cuda();
-    common_dtype_ = std::get<1>(common_type);
-    may_have_differing_types = !std::get<2>(common_type);
-    bool has_cpu_scalar = false;
+  // The computation dtype cannot be inferred if there are different input dtypes
+  //   and inputs are not promoted, so the dtype of undefined outputs and
+  //   how to cast to outputs cannot be determined.
+  TORCH_INTERNAL_ASSERT(!(has_different_input_dtypes && !promote_inputs_to_common_dtype_ &&
+                        (has_undefined_outputs || enforce_safe_casting_to_output_ ||
+                        cast_common_dtype_to_outputs_)));
+
+  // Computes a common dtype, if needed
+  // NOTE: only occurs if there are multiple distinct dtypes
+  if (has_different_input_dtypes && promote_inputs_to_common_dtype_) {
+    common_dtype_ = compute_common_dtype();
+  }
+
+  // Reviews operands (2/2) - CUDA and CPU cases
```

Comment:
I'm still trying to understand the new code, but I'm not a big fan of having copy pasted the code for the cpu / cuda cases (with the behavior changes you described, ofc)

Reply:
Me either. It may be a reactionary response to the previous code, which was very hard to parse with a plethora of cases. The actual lines of overlap is smaller than it appears. I'll take another look at merging them and if it doesn't impact the readability much I'll do it.